### PR TITLE
Restructure project

### PR DESCRIPTION
**Pull request by kalaninja**
- merged `types` and `rpc-types`
- moved `storage` to `pallet`
- moved all crypto deps to `crypto`
- moved 'rpc-sections' and related types to `rpc-api`
- reduced number of subprojects

I propose the following structure (layers) in order of dependency:
- common (will now contain basic type features e.g. unions, tuples, unit, result, etc.)
- crypto, transport, scale
- rpc (codegen + api)
- pallet
- api

### CODE REVIEW

**Comment by vnabiev** [comment]
I like your approach of reducing the number of packages and the structure in general, however I would propose to make some changes:
1. to rename the package `types` to something like `rpc-api`;
2. to add subpackages `types`, `readers`, `writers`, `decoders`, `encoders` into `rpc-api` and then spread the entities of the package across these ones properly;
3. to add subpackage `sections` and move back from `api` the following classes `State`, `System`, `Author`, `Chain`;
4. to remove class `StateProvider`

So we will have everything related to the RPC's API in the same place and the "pallet" package will keep dependency of "rpc-api" that makes sense, especially when we will implement the other pallet's APIs.

**Comment by kalaninja** [comment]
Agree. made the changes. Also, updated the commit message.

**Review by vnabiev — APPROVED** [review]

File: pallet/pallet-codegen/src/main/java/com/strategyobject/substrateclient/pallet/codegen/PalletAnnotatedInterface.java [hunk 1]
```diff
@@ -54,11 +55,11 @@ public void generateClass(ProcessorContext context) throws ProcessingException,
         val typeSpecBuilder = TypeSpec.classBuilder(className)
                 .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                 .addSuperinterface(TypeName.get(interfaceElement.asType()))
-                .addField(Rpc.class, RPC, Modifier.FINAL, Modifier.PRIVATE);
+                .addField(State.class, STATE, Modifier.FINAL, Modifier.PRIVATE);
 
         val constructorBuilder = createConstructorBuilder();
 
-        for (val method : interfaceElement.getEnclosedElements()) {
+        for (Element method : interfaceElement.getEnclosedElements()) {
```
[the remark this answers is not in the document]
Comment (kalaninja) [reply pallet/pallet-codegen/src/main/java/com/strategyobject/substrateclient/pallet/codegen/PalletAnnotatedInterface.java:62]: to help sonar derive type properly

File: pallet/src/test/java/com/strategyobject/substrateclient/pallet/storage/Chain.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package com.strategyobject.substrateclient.pallet.storage;
```
[the remark this answers is not in the document]
Comment (kalaninja) [reply pallet/src/test/java/com/strategyobject/substrateclient/pallet/storage/Chain.java:1]: yes, it is not needed now. So, removed it.

File: rpc/rpc-api/src/test/java/com/strategyobject/substrateclient/rpc/api/section/StateTests.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    @Test
+    void getKeys() throws Exception {
+        try (val wsProvider = connect()) {
+            val state = RpcGeneratedSectionFactory.create(State.class, wsProvider);
+
+            // xxhash128("Balances") = 0xc2261276cc9d1f8598ea4b6a74b15c2f
+            // xxhash128("StorageVersion") = 0x308ce9615de0775a82f8a94dc3d285a1
+            val key = "0xc2261276cc9d1f8598ea4b6a74b15c2f308ce9615de0775a82f8a94dc3d285a1"; // TODO implement and use `xxhash`
```
[the remark this answers is not in the document]
Comment (kalaninja) [reply rpc/rpc-api/src/test/java/com/strategyobject/substrateclient/rpc/api/section/StateTests.java:61]: no problem, removed it

**Pull request merged**